### ISSUE TITLE
`ICE` found in `compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:179:90`

### ISSUE THREAD

**Issue by Snowapril** [rust-lang/rust-clippy#8830]
### Summary

While working on this branch(https://github.com/snowapril/rustpython/tree/impl_contextvar commit(ca956fc16c497b4b7b7ccce534068569c3ce1266)), I met ICE issue.
Here is full output of clippy.
```bash
> RUST_BACKTRACE=full cargo clippy
    Checking rustpython-vm v0.1.2 (/home/snowapril/RustPython/vm)
warning: unknown lint: `clippy::needless_match`
   --> vm/src/builtins/function.rs:512:17
    |
512 |         #[allow(clippy::needless_match)] // False positive on nightly
    |                 ^^^^^^^^^^^^^^^^^^^^^^ help: did you mean: `clippy::needless_bool`
    |
    = note: `#[warn(unknown_lints)]` on by default

error: internal compiler error: compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:179:90: Failed to normalize <std::option::Option<&std::option::Option<object::core::PyRef<builtins::pystr::PyStr>>> as std::iter::Iterator>::Item, maybe try to call `try_normalize_erasing_regions` instead
```

### Version

```text
rustc 1.60.0 (7737e0b5c 2022-04-04)
binary: rustc
commit-hash: 7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c
commit-date: 2022-04-04
host: x86_64-unknown-linux-gnu
release: 1.60.0
LLVM version: 14.0.0
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
  thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1160:9
stack backtrace:
   0:     0x7f607b47586d - std::backtrace_rs::backtrace::libunwind::trace::hee598835bc88d35b
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f607b47586d - std::backtrace_rs::backtrace::trace_unsynchronized::h9cdc730ba5cf5d72
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f607b47586d - std::sys_common::backtrace::_print_fmt::h75aeaf7ed30e43fa
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f607b47586d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h606862f787600875
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f607b4d54ec - core::fmt::write::he803f0f418caf762
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/fmt/mod.rs:1190:17
   5:     0x7f607b465a98 - std::io::Write::write_fmt::h70bc45872f37e7bb
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/io/mod.rs:1657:15
   6:     0x7f607b479847 - std::sys_common::backtrace::_print::h64d038cf8ac3e13e
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f607b479847 - std::sys_common::backtrace::print::h359300b4a7fccf65
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f607b479847 - std::panicking::default_hook::{{closure}}::hf51be35e2f510149
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:295:22
   9:     0x7f607b479510 - std::panicking::default_hook::h03ca0f22e1d2d25e
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:314:9
  10:     0x56001af1fd61 - clippy_driver[d588fef56f57e4ae]::ICE_HOOK::{closure#0}::{closure#0}
  11:     0x7f6065fdce43 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h96a12586f174161c
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/alloc/src/boxed.rs:1867:9
  12:     0x7f6065feb4ac - proc_macro::bridge::client::<impl proc_macro::bridge::Bridge>::enter::{{closure}}::{{closure}}::h377e7953991447d5
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/proc_macro/src/bridge/client.rs:319:21
  13:     0x7f6065f57f60 - std::panicking::update_hook::{{closure}}::h8adb07778c9252b7
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:258:41
  14:     0x7f607b47a124 - std::panicking::rust_panic_with_hook::h3b7380e99b825b63
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:702:17
  15:     0x7f607ccdb301 - std[cd3cf6af28dff6de]::panicking::begin_panic::<rustc_errors[bd800fb2a5e0b9f5]::ExplicitBug>::{closure#0}
  16:     0x7f607ccdb266 - std[cd3cf6af28dff6de]::sys_common::backtrace::__rust_end_short_backtrace::<std[cd3cf6af28dff6de]::panicking::begin_panic<rustc_errors[bd800fb2a5e0b9f5]::ExplicitBug>::{closure#0}, !>
  17:     0x7f607cce163f - std[cd3cf6af28dff6de]::panicking::begin_panic::<rustc_errors[bd800fb2a5e0b9f5]::ExplicitBug>
  18:     0x7f607ccee2ed - std[cd3cf6af28dff6de]::panic::panic_any::<rustc_errors[bd800fb2a5e0b9f5]::ExplicitBug>
  19:     0x7f607cced5a7 - <rustc_errors[bd800fb2a5e0b9f5]::HandlerInner>::bug
  20:     0x7f607cced0d0 - <rustc_errors[bd800fb2a5e0b9f5]::Handler>::bug
  21:     0x7f607cb6f609 - rustc_middle[c145adf1c4567bf]::ty::context::tls::with_opt::<rustc_middle[c145adf1c4567bf]::util::bug::opt_span_bug_fmt<rustc_span[4b4bb4534443c42f]::span_encoding::Span>::{closure#0}, ()>
  22:     0x7f607cb6fff0 - rustc_middle[c145adf1c4567bf]::util::bug::opt_span_bug_fmt::<rustc_span[4b4bb4534443c42f]::span_encoding::Span>
  23:     0x7f607cb6ff66 - rustc_middle[c145adf1c4567bf]::util::bug::bug_fmt
  24:     0x7f607d9e111a - <rustc_middle[c145adf1c4567bf]::ty::normalize_erasing_regions::NormalizeAfterErasingRegionsFolder as rustc_middle[c145adf1c4567bf]::ty::fold::TypeFolder>::fold_ty
  25:     0x56001b279ea1 - clippy_utils[fada9aa31059ad06]::ty::get_associated_type
  26:     0x56001afe1fc8 - clippy_lints[a93cda6d98502404]::methods::iter_overeager_cloned::check
  27:     0x56001b17de1f - <clippy_lints[a93cda6d98502404]::methods::Methods as rustc_lint[fc067b814fceecd7]::passes::LateLintPass>::check_expr
  28:     0x7f607c9a7034 - <rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects as rustc_lint[fc067b814fceecd7]::passes::LateLintPass>::check_expr
  29:     0x7f607bca24d0 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_expr
  30:     0x7f607bca22b0 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_nested_body
  31:     0x7f607bca26c2 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_fn
  32:     0x7f607bca592b - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_expr::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  33:     0x7f607bca24db - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_expr
  34:     0x7f607bca5a6d - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_expr::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  35:     0x7f607bca24db - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_expr
  36:     0x7f607bca2de9 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_block
  37:     0x7f607bca24db - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_expr
  38:     0x7f607bca22b0 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_nested_body
  39:     0x7f607bca26c2 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_fn
  40:     0x7f607bca5221 - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_impl_item::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  41:     0x7f607bca1a3c - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_nested_impl_item
  42:     0x7f607bca5418 - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_impl_item_ref::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  43:     0x7f607bca3c8f - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_item::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  44:     0x7f607bca1576 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_nested_item
  45:     0x7f607bca3627 - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_mod::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  46:     0x7f607bca2b5c - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_mod
  47:     0x7f607bca1576 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_nested_item
  48:     0x7f607bca3627 - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_mod::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  49:     0x7f607bca2b5c - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_mod
  50:     0x7f607bca1576 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_nested_item
  51:     0x7f607bca3627 - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_mod::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  52:     0x7f607bca2b5c - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_mod
  53:     0x7f607bca1576 - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_nested_item
  54:     0x7f607bca3627 - rustc_hir[cb2e3a51e91fc9ed]::intravisit::walk_mod::<rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>>
  55:     0x7f607bca2b5c - <rustc_lint[fc067b814fceecd7]::late::LateContextAndPass<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects> as rustc_hir[cb2e3a51e91fc9ed]::intravisit::Visitor>::visit_mod
  56:     0x7f607bca3330 - rustc_lint[fc067b814fceecd7]::late::late_lint_pass_crate::<rustc_lint[fc067b814fceecd7]::late::LateLintPassObjects>
  57:     0x7f607da73036 - rustc_lint[fc067b814fceecd7]::late::late_lint_crate::<rustc_lint[fc067b814fceecd7]::BuiltinCombinedLateLintPass>
  58:     0x7f607daa178b - <rustc_session[95a8f1c94d15e5dc]::session::Session>::time::<(), rustc_lint[fc067b814fceecd7]::late::check_crate<rustc_lint[fc067b814fceecd7]::BuiltinCombinedLateLintPass, rustc_interface[afc566840a941718]::passes::analysis::{closure#5}::{closure#0}::{closure#3}::{closure#0}::{closure#0}>::{closure#0}::{closure#0}>
  59:     0x7f607daa4926 - <core[2734bb1f7d9ec6a5]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[afc566840a941718]::passes::analysis::{closure#5}::{closure#0}::{closure#3}> as core[2734bb1f7d9ec6a5]::ops::function::FnOnce<()>>::call_once
  60:     0x7f607daa6b21 - <core[2734bb1f7d9ec6a5]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[afc566840a941718]::passes::analysis::{closure#5}::{closure#0}> as core[2734bb1f7d9ec6a5]::ops::function::FnOnce<()>>::call_once
  61:     0x7f607daa30a8 - <rustc_session[95a8f1c94d15e5dc]::session::Session>::time::<(), rustc_interface[afc566840a941718]::passes::analysis::{closure#5}>
  62:     0x7f607da87f6c - rustc_interface[afc566840a941718]::passes::analysis
  63:     0x7f607dfdfb62 - <rustc_query_system[cf3d2502706fe991]::dep_graph::graph::DepGraph<rustc_middle[c145adf1c4567bf]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[c145adf1c4567bf]::ty::context::TyCtxt, (), core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>
  64:     0x7f607e032303 - rustc_data_structures[762b64c8eff0c924]::stack::ensure_sufficient_stack::<(core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>, rustc_query_system[cf3d2502706fe991]::dep_graph::graph::DepNodeIndex), rustc_query_system[cf3d2502706fe991]::query::plumbing::execute_job<rustc_query_impl[8218de0923c6cd45]::plumbing::QueryCtxt, (), core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>::{closure#3}>
  65:     0x7f607df21bae - rustc_query_system[cf3d2502706fe991]::query::plumbing::try_execute_query::<rustc_query_impl[8218de0923c6cd45]::plumbing::QueryCtxt, rustc_query_system[cf3d2502706fe991]::query::caches::DefaultCache<(), core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>>
  66:     0x7f607df60c05 - rustc_query_system[cf3d2502706fe991]::query::plumbing::get_query::<rustc_query_impl[8218de0923c6cd45]::queries::analysis, rustc_query_impl[8218de0923c6cd45]::plumbing::QueryCtxt>
  67:     0x7f607da66608 - <rustc_interface[afc566840a941718]::passes::QueryContext>::enter::<rustc_driver[6eeb127b85fe99ac]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>
  68:     0x7f607da54ac5 - rustc_interface[afc566840a941718]::interface::create_compiler_and_run::<core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>, rustc_driver[6eeb127b85fe99ac]::run_compiler::{closure#1}>
  69:     0x7f607da3a134 - std[cd3cf6af28dff6de]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[afc566840a941718]::util::run_in_thread_pool_with_globals<rustc_interface[afc566840a941718]::interface::run_compiler<core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>, rustc_driver[6eeb127b85fe99ac]::run_compiler::{closure#1}>::{closure#0}, core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>::{closure#0}, core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>
  70:     0x7f607da6d3e9 - <<std[cd3cf6af28dff6de]::thread::Builder>::spawn_unchecked_<rustc_interface[afc566840a941718]::util::run_in_thread_pool_with_globals<rustc_interface[afc566840a941718]::interface::run_compiler<core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>, rustc_driver[6eeb127b85fe99ac]::run_compiler::{closure#1}>::{closure#0}, core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>::{closure#0}, core[2734bb1f7d9ec6a5]::result::Result<(), rustc_errors[bd800fb2a5e0b9f5]::ErrorReported>>::{closure#1} as core[2734bb1f7d9ec6a5]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  71:     0x7f607b485d03 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hf70ac038171e3e1a
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/alloc/src/boxed.rs:1853:9
  72:     0x7f607b485d03 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he6690128792365ad
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/alloc/src/boxed.rs:1853:9
  73:     0x7f607b485d03 - std::sys::unix::thread::Thread::new::thread_start::ha07928d93d5a5ec9
                               at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys/unix/thread.rs:108:17
  74:     0x7f607b3a0609 - start_thread
  75:     0x7f607b176293 - clone
  76:                0x0 - <unknown>
  ```

  </p>
</details>

**Comment by matthiaskrgr** [comment]
sounds like a rustc bug.
This probably also ICEs with `-Zmir-opt-level=3 --emit=mir` or something like that.

**Comment by giraffate** [comment]
It seems to be a duplicate of https://github.com/rust-lang/rust-clippy/issues/8662, so I'm closing this.
